mypy/fastparse.py

```diff
@@ -216,19 +216,26 @@ def parse(
     try:
         ignore_comment_errors = options and options.ignore_comment_errors
 
-        def try_parse_ast(source: str) -> tuple[AST | None, str]:
+        def try_parse_ast(source: str | bytes) -> tuple[AST | None, str | bytes]:
             try:
                 ast = ast3_parse(source, fnam, "exec", feature_version=feature_version)
             except SyntaxError as e:
 
                 def strip_comment(line: str) -> str:
                     return re.sub(r"([^#]*).*(\r\n|\r|\n)?", r"\1\2", line)
 
-                def strip_comment_from_line(text: str, target_lineno: int) -> str:
-                    lines = text.splitlines(keepends=True)
-                    return "".join(
-                        line if lineno != target_lineno else strip_comment(line)
-                        for lineno, line in enumerate(lines, start=1)
+                def strip_commentb(line: bytes) -> bytes:
+                    return re.sub(rb"([^#]*).*(\r\n|\r|\n)?", rb"\1\2", line)
+
+                def strip_comment_from_line(text: str | bytes, target_lineno: int) -> str | bytes:
+                    if isinstance(text, str):
+                        return "".join(
+                            line if lineno != target_lineno else strip_comment(line)
+                            for lineno, line in enumerate(text.splitlines(keepends=True), start=1)
+                        )
+                    return b"".join(
+                        line if lineno != target_lineno else strip_commentb(line)
+                        for lineno, line in enumerate(text.splitlines(keepends=True), start=1)
```

Comment:
I'd probably rather do that branching inside `strip_comment(line: str | bytes)` since you end up duplicating less code this way :)

Reply:
Yes, this situation is annoying. I tried using `AnyStr`, but couldn't get it to work. I think we may have an issue where you get an unbound error if you do:
```python
s_or_b: str | bytes = "a"
any_str: AnyStr = s_or_b  # AnyStr is unbound error, which doesn't make sense b/c it's constrained
```
I don't think I can avoid duplicating the `join` because in on instance I join on `""`, and  in another `b""`. I tried something like `empty = "" if intanceof(str, text) else b""`, then doing `empty.join(...)`, but couldn't get the types to work out.

Reply:
`type(str)()` maybe? :P